lib/Sema/NameBinding.cpp

```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
+        auto nameDecl = cast<UnresolvedDeclRefExpr>(PUE->getFn());
+        if (nameDecl->getName().getBaseName().str() != "!") {
+          return false;
+        }
+        return !evaluateCondition(PUE->getArg());
+      }
+
+      // Evaluate a target config call expression.
+      if (auto *CE = dyn_cast<CallExpr>(condition)) {
+        // Get the arg, which should be in a paren expression.
+        auto fnNameExpr = dyn_cast<UnresolvedDeclRefExpr>(CE->getFn());
+        if (!fnNameExpr) {
+          return false;
+        }
+        auto fnName = fnNameExpr->getName().getBaseName().str();
+        auto *PE = dyn_cast<ParenExpr>(CE->getArg());
+        if (!PE) {
+          return false;
+        }
+        if (fnName.equals("_compiler_version")) {
+          auto SLE = cast<StringLiteralExpr>(PE->getSubExpr());
+
+          // Parse the compiler version string.  Diagnostics have already been
+          // emitted so suppress them.
+          DiagnosticTransaction trans(Binder.Context.Diags);
+          auto versionRequirement =
+            version::Version::parseCompilerVersionString(SLE->getValue(),
+                                                         SLE->getLoc(),
+                                                         &Binder.Context.Diags);
+          trans.abort();
+
+          auto thisVersion = version::Version::getCurrentCompilerVersion();
+          auto VersionNewEnough = thisVersion >= versionRequirement;
+          return VersionNewEnough;
+        } else if (fnName.equals("swift")) {
+          auto PUE = dyn_cast<PrefixUnaryExpr>(PE->getSubExpr());
+          if (!PUE) {
+            return false;
+          }
+          auto prefix = dyn_cast<UnresolvedDeclRefExpr>(PUE->getFn());
+          auto versionArg = PUE->getArg();
+          auto versionStartLoc = versionArg->getStartLoc();
+          auto endLoc = Lexer::getLocForEndOfToken(Context.SourceMgr,
+                                                   versionArg->getSourceRange().End);
+          CharSourceRange versionCharRange(Context.SourceMgr, versionStartLoc,
+                                           endLoc);
+          auto versionString = Context.SourceMgr.extractText(versionCharRange);
+
+          // Parse the version string.  Diagnostics have already been emitted
+          // so suppress them.
+          DiagnosticTransaction trans(Binder.Context.Diags);
+          auto versionRequirement =
+            version::Version::parseVersionString(versionString,
+                                                 versionStartLoc,
+                                                 &Binder.Context.Diags);
+          trans.abort();
+
+          assert(versionRequirement.hasValue());
+          auto thisVersion = Context.LangOpts.EffectiveLanguageVersion;
+          assert(prefix->getName().getBaseName().str().equals(">="));
+
+          auto VersionNewEnough = thisVersion >= versionRequirement.getValue();
+          return VersionNewEnough;
```

Comment:
I was afraid of this.  It might be useful to have the new stripped-down `classify` function set a trinary value indicating a clause was evaluated then and there and assert down here that we've already handled this.